py/server/deephaven/barrage.py

```diff
@@ -0,0 +1,208 @@
+            authenticator, it must conform to the specific requirement of the authenticator
+        use_tls (bool): if True, use a TLS connection.  Defaults to False
+        tls_root_certs (bytes): PEM encoded root certificates to use for TLS connection, or None to use system defaults.
+             If not None implies use a TLS connection and the use_tls argument should have been passed
+             as True. Defaults to None
+
+    Returns:
+        a Deephaven Barrage session
+
+    Raises:
+        DHError
+    """
+    try:
+        if tls_root_certs and not use_tls:
+            raise DHError(message="tls_root_certs is provided but use_tls is False")
+
+        target_uri = f"{host}:{port}"
+        if use_tls:
+            target_uri = f"dh://{target_uri}"
+        else:
+            target_uri = f"dh+plain://{target_uri}"
+
+        j_client_config = _build_client_config(target_uri, tls_root_certs)
+        auth = f"{auth_type} {auth_token}"
+
+        try:
+            return _get_barrage_session_via_api_server(j_client_config, auth)
+        except:
+            # fall back to the direct way when we don't have a fully initialized server, used for testing only
+            # TODO: remove when we are done with restructuring the integrations tests wiring
+            #  https://github.com/deephaven/deephaven-core/issues/5401
+            return _get_barrage_session_direct(j_client_config, auth)
+    except Exception as e:
+        raise DHError(e, "failed to get a barrage session to the target remote Deephaven server.") from e
+
+
+def _get_barrage_session_via_api_server(client_config: jpy.JType, auth: str) -> BarrageSession:
+    j_barrage_session_factory_creator = _JDeephavenApiServer.getInstance().sessionFactoryCreator()
+    j_barrage_session_factory = j_barrage_session_factory_creator.barrageFactory(client_config)
+    j_managed_channel = j_barrage_session_factory.managedChannel()
+    if auth:
+        j_session_config = (_JSessionConfig.builder()
+                            .authenticationTypeAndValue(auth)
+                            .build())
+        j_barrage_session = j_barrage_session_factory.newBarrageSession(j_session_config)
+    else:
+        j_barrage_session = j_barrage_session_factory.newBarrageSession()
+    return BarrageSession(j_barrage_session, j_managed_channel)
+
+def _get_barrage_session_direct(client_config: jpy.JType, auth: str) -> BarrageSession:
+    """Note, this is used for testing only. This way of constructing a Barrage session is less efficient because it does
+    not share any of the state or configuration that the server provides; namely, when you are doing it with the server
+    context it provides a singleton executor, allocator, outbound SSL configuration, and the ability for the server to
+    hook in additional channel building options.
+
+    TODO: remove when we are done with restructuring the integrations tests wiring
+    https://github.com/deephaven/deephaven-core/issues/5401.
```

Comment:
Should this be in the pydoc or as a normal comment.  Does pycharm pick up todos in pydocs?

Reply:
Sphinx won't pick up docstrings on private functions.